src/lsm/scan_tree.zig

```diff
@@ -581,6 +701,56 @@ fn ScanTreeLevelType(comptime ScanTree: type, comptime Storage: type) type {
             }
         }
 
+        pub fn probe(self: *ScanTreeLevel, key: Key) void {
+            maybe(self.manifest == .iterating_tables or
+                self.manifest == .iterating_blocks or
+                self.manifest == .finished);
+
+            switch (self.values) {
+                .fetching => {},
+                .buffered => |buffer| {
+                    assert(buffer.len > 0);
+                    const slice: []const Value = ScanTree.probe_values(
+                        self.scan.direction,
+                        buffer,
+                        key,
+                    );
+
+                    if (slice.len == 0) {
+                        self.values = .fetching;
+                    } else {
+                        if (self.manifest == .iterating_blocks) {
+                            // The next exclusive key must be ahead of (or equals) the probe key,
+                            // so the level iterator state can be preserved without reading the
+                            // index block again.
+                            const key_exclusive_next = self.manifest.iterating_blocks;
+                            assert(switch (self.scan.direction) {
+                                .ascending => key_exclusive_next >= key,
+                                .descending => key_exclusive_next <= key,
+                            });
+                        }
+
+                        self.values = .{
+                            .buffered = slice,
+                        };
+                    }
+                },
+                .finished => return,
+            }
+
+            if (self.values == .fetching) {
+                // The key couldn't be found in the buffered data.
+                // The level iterator must read the index block again from the new key range.
+                //
+                // TODO: We may use the already buffered `index_block` to check if the key
+                // is present in other value blocks within the same table, advancing the level
+                // iterator instead of calling `table_next(.begin)`.
+                // However, it's most likely the index block is still in the grid cache, so this
+                // may not represent any real improvement.
```

Comment:
Yes, either `.finished` or `.iterating_blocks`, it will need to read the index block again to determine if the probe_key can be in a value block ahead of the current one. The `TODO` comment applies to both cases.